stan/math/opencl/kernel_generator/unary_function_cl.hpp

```diff
@@ -0,0 +1,152 @@
+#ifndef STAN_MATH_OPENCL_KERNEL_GENERATOR_UNARY_FUNCTION_CL_HPP
+#define STAN_MATH_OPENCL_KERNEL_GENERATOR_UNARY_FUNCTION_CL_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/matrix_cl_view.hpp>
+#include <stan/math/opencl/err/check_matching_dims.hpp>
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/opencl/kernel_generator/type_str.hpp>
+#include <stan/math/opencl/kernel_generator/name_generator.hpp>
+#include <stan/math/opencl/kernel_generator/operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/as_operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/is_valid_expression.hpp>
+#include <string>
+#include <type_traits>
+#include <set>
+#include <utility>
+
+namespace stan {
+namespace math {
+
+/**
+ * Represents a unary function in kernel generator expressions.
+ * @tparam Derived derived type
+ * @tparam T type of argument
+ */
+template <typename Derived, typename T>
+class unary_function_cl
+    : public operation_cl<
+          Derived, typename std::remove_reference_t<T>::ReturnScalar, T> {
+ public:
+  using ReturnScalar = typename std::remove_reference_t<T>::ReturnScalar;
+  static_assert(std::is_floating_point<ReturnScalar>::value,
+                "unary_function_cl: argument must be expression with floating "
+                "point return type!");
+  using base = operation_cl<Derived, ReturnScalar, T>;
+  using base::var_name;
+
+  /**
+   * Constructor
+   * @param a argument expression
+   * @param fun function
+   */
+  unary_function_cl(T&& a, const std::string& fun)
+      : base(std::forward<T>(a)), fun_(fun) {}
+
+  /**
+   * generates kernel code for this expression.
+   * @param i row index variable name
+   * @param j column index variable name
+   * @param var_name_arg variable name of the nested expression
+   * @return part of kernel with code for this expression
+   */
+  inline kernel_parts generate(const std::string& i, const std::string& j,
+                               const std::string& var_name_arg) const {
+    kernel_parts res{};
+    res.body = type_str<ReturnScalar>() + " " + var_name + " = " + fun_ + "("
+               + var_name_arg + ");\n";
+    return res;
+  }
+
+  /**
+   * View of a matrix that would be the result of evaluating this expression.
+   * @return view
+   */
+  inline matrix_cl_view view() const {
+    return std::get<0>(base::arguments_).view();
+  }
+
+ protected:
+  std::string fun_;
+};
+
+/**
+ * generates a class and function for an unary function.
+ * @param fun function
+ */
+#define ADD_UNARY_FUNCTION(fun)                                                \
+  template <typename T>                                                        \
+  class fun##__ : public unary_function_cl<fun##__<T>, T> {                    \
```

Comment:
this comes out to `sin__` right? C++ reserves anything with double underscores in it for the compiler

Reply:
Faixed.